pkg/activator/throttler.go

```diff
@@ -0,0 +1,212 @@
+	OverloadMessage = "activator overload"
+)
+
+// ThrottlerParams defines the parameters of the Throttler.
+type ThrottlerParams struct {
+	BreakerParams queue.BreakerParams
+	Logger        *zap.SugaredLogger
+	GetEndpoints  func(RevisionID) (int32, error)
+	GetRevision   func(RevisionID) (*v1alpha1.Revision, bool, error)
+}
+
+// NewThrottler creates a new Throttler.
+func NewThrottler(params ThrottlerParams) *Throttler {
+	breakers := make(map[RevisionID]*queue.Breaker)
+	return &Throttler{breakers: breakers, breakerParams: params.BreakerParams, logger: params.Logger, getEndpoints: params.GetEndpoints, getRevision: params.GetRevision}
+}
+
+// Throttler keeps the mapping of Revisions to Breakers
+// and allows updating max concurrency dynamically of respective Breakers.
+// Max concurrency is essentially the number of semaphore tokens the Breaker has in rotation.
+// The manipulation of the parameter is done via `UpdateCapacity()` method.
+// It enables the use case to start with max concurrency set to 0 (no requests are sent because no endpoints are available)
+// and gradually increase its value depending on the external condition (e.g. new endpoints become available)
+type Throttler struct {
+	breakers      map[RevisionID]*queue.Breaker
+	breakerParams queue.BreakerParams
+	logger        *zap.SugaredLogger
+	getEndpoints  func(RevisionID) (int32, error)
+	getRevision   func(RevisionID) (*v1alpha1.Revision, bool, error)
+	mux           sync.Mutex
+}
+
+// Remove deletes the breaker from the bookkeeping.
+func (t *Throttler) Remove(rev RevisionID) {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	delete(t.breakers, rev)
+}
+
+// UpdateCapacity updates the max concurrency of the Breaker corresponding to a revision.
+func (t *Throttler) UpdateCapacity(rev RevisionID, size int32) error {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	revision, err := t.revision(rev)
+	if err != nil {
+		return err
+	}
+	breaker, _ := t.getOrCreateBreaker(rev)
+	err = t.updateCapacity(revision, breaker, size)
+	return err
+}
+
+// Try potentially registers a new breaker in our bookkeeping
+// and executes the `function` on the Breaker.
+// It returns an error if either breaker doesn't have enough capacity,
+// or breaker's registration didn't succeed, e.g. getting endpoints or update capacity failed.
+func (t *Throttler) Try(rev RevisionID, function func()) error {
+	t.mux.Lock()
+	breaker, existed := t.getOrCreateBreaker(rev)
+	if !existed {
+		err := t.forceUpdateCapacity(rev, breaker)
+		if err != nil {
+			return err
+		}
+	}
+	t.mux.Unlock()
+	if ok := breaker.Maybe(function); !ok {
+		return errors.New(OverloadMessage)
+	}
+	return nil
+}
+
+// This method updates Breaker's concurrency and requires external synchronization, `updateCapacity` requires `mux` to be held.
+func (t *Throttler) updateCapacity(revision *v1alpha1.Revision, breaker *queue.Breaker, size int32) (err error) {
+	cc := int32(revision.Spec.ContainerConcurrency)
```

Comment:
This is the only bit of the revision passed into `updateCapacity`. Should we pass `ContainerConcurrency` instead to avoid passing the revision around?

Reply:
Since we are passing the pointer to the object, it is the same as if we pass the int32. So there is no disadvantage. At the same time we only need to cast to int32 once, so I'm more in favor of the current approach.

Reply:
This is not about overhead of passing the object but rather API surface wise it's clearer if you're passing a narrower set of parameters. Either way is fine.

Reply:
I see your point, for me it is more about casting to int32 in a single place, so if we need to change the method, we do it once.